### PR TITLE
Fix bug when using an private aliased factory service

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17910, #17915 |
| License | MIT |
| Doc PR | - |

/cc @xabbuh

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php [hunk 1]
```diff
@@ -131,6 +131,6 @@ private function updateFactoryServiceReference($factoryService, $currentId, $new
             return;
         }
 
-        return $currentId === $factoryService ? $newId : $currentId;
+        return $currentId === $factoryService ? $newId : $factoryService;
```
Comment (WouterJ) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: `$currentId` is the current ID of the removed alias, while `$factoryService` is the original ID of the factory service.

If the factory service equals the ID of the removed alias, it should be updated to become the actual service ID. However, if the factory service does not equal the ID of the removed alias, the original value should be used.

With the old code, it was always replacing the original code, either with the actual definition ID or the ID of the removed alias (which was causing the troubles people were reporting).

**Comment by xabbuh** [comment]
:+1: 

Oh my God, what a silly mistake. Thanks Wouter for the fix!

Status: Reviewed

**Comment by iltar** [comment]
Was the test failing before? If not, a test would be nice to prevent this bug from happening again

**Comment by WouterJ** [comment]
@iltar I've added a test now (although this is a very specific test, so I'm not 100% sure if we should include it).

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ReplaceAliasByActualDefinitionPassTest.php [hunk 2]
```diff
@@ -45,6 +47,26 @@ public function testProcess()
     }
 
     /**
+     * @group legacy
```
Comment (WouterJ) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ReplaceAliasByActualDefinitionPassTest.php]: it's not really legacy in 2.3, but this makes things easier to merge to 2.7/2.8

**Comment by fabpot** [comment]
Thank you @WouterJ.

File: src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php (hunk 1 above)
Comment (pborreli) [inline src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php]: good catch @WouterJ :clap:

**Pull request closed without merging**